### PR TITLE
EventMachine -> concurrent-ruby, take two

### PR DESCRIPTION
**Pull request by matthewd**

### CODE REVIEW

**Comment by dhh** [comment]
Tested with Basecamp 👍

**Comment by dhh** [comment]
So feel free to merge.

**Comment by TiagoCardoso1983** [comment]
@matthewd nio4r uses libev for MRI. If the purpose of removing eventmachine was to have support for windows MRI, that won't do it, at least until someone port nio4r to libuv.

File: actioncable/lib/action_cable/connection/client_socket.rb [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    # Copyright (c) 2010-2015 James Coglan
+    class ClientSocket # :nodoc:
+      def self.determine_url(env)
+        scheme = secure_request?(env) ? 'wss:' : 'ws:'
+        "#{ scheme }//#{ env['HTTP_HOST'] }#{ env['REQUEST_URI'] }"
+      end
+
+      def self.secure_request?(env)
+        return true if env['HTTPS'] == 'on'
+        return true if env['HTTP_X_FORWARDED_SSL'] == 'on'
```
Comment (ianks) [inline actioncable/lib/action_cable/connection/client_socket.rb]: Might be wise to use Rack::Reqest#ssl? here.

File: actioncable/lib/action_cable/server/base.rb [hunk 2]
```diff
@@ -29,25 +36,31 @@ def disconnect(identifiers)
 
       # Gateway to RemoteConnections. See that class for details.
       def remote_connections
-        @remote_connections ||= RemoteConnections.new(self)
+        @remote_connections || @mutex.synchronize { @remote_connections ||= RemoteConnections.new(self) }
```
Comment (ianks) [inline actioncable/lib/action_cable/server/base.rb]: Why not use AtomicReferences for these?

**Comment by matthewd** [comment]
> nio4r uses libev for MRI

@TiagoCardoso1983 [except on Windows](https://github.com/celluloid/nio4r/blob/master/lib/nio.rb#L16), where it falls back to `Kernel#select`.

File: actioncable/lib/action_cable/connection/client_socket.rb (hunk 1 above)
Comment (exAspArk) [inline actioncable/lib/action_cable/connection/client_socket.rb]: The same with the line above, maybe it's better to use `Rack::Reqest#scheme == 'https'`?

**Comment by TiagoCardoso1983** [comment]
@matthewd what about https://github.com/celluloid/nio4r/issues/52? nio4r pure ruby implementation is quite untested and contains recursive locking issues.

**Comment by matthewd** [comment]
@TiagoCardoso1983 we're not using the block form of `select`, so that bug doesn't affect us. But even if it did, the existence of bugs in a library doesn't seem particularly shocking to me. They will exist, and they will hopefully be fixed.

I'm not sure what you're implying we should do differently here. Do you have a specific recommendation?

**Comment by TiagoCardoso1983** [comment]
@matthewd the main reason eventmachine was removed is because it doesn't build under windows, right? I do prefer the implementation using nio4r, just wanted to state the windows potential issue while rails is still in beta mode so anyone with access to the environment could test it. Rails historically refrains from using dependencies with extensions so that it can "go where ruby goes". This contract has been broken once already with the html sanitizer gem. I'm afraid most of the tester pool will be using some *NIX form running against the libev reactor and will not test the fallback IO.select, so just wanted to state the potential drawback before the official release of 5.

**Comment by schmurfy** [comment]
seriously is there really anyone running production rails code under windows ? For development I can understand but I would be impressed otherwise. Eventmachine also has had bugs of its own for a while now althought development seems to now happens again on it after years of a mostly abandonned codebase with a laundry of pull requests ready to be merged with nobody to do it.

concurrent-ruby and nio4r seems like the best bet for me and a really good choice.

**Comment by dhh** [comment]
I'm not concerned about Windows production deployments either. The main use
case is making sure it's OK for development. What we have now works for
that. If someone finds bugs during Windows development, we can treat them
as any other type of bug.

**Comment by TiagoCardoso1983** [comment]
@schmurfy I don't think anyone wants to deploy action cable in production relying on IO.select, whatever the platform is :) Windows is still a relevant development environment, and is annoying mostly because a lot of dependencies go the "we only support UNIX" way. 

Anyway, I can see this going the activejob/rails way in that one supports the basic setup for synch-jobs/webrick and then it's up for the developer to decide the production setup and dependencies. Main requirement for the basic setup is that it works well where ruby works.

**Comment by SuShu19** [comment]
> seriously is there really anyone running production rails code under windows ? For development I can understand but I would be impressed otherwise. Eventmachine also has had bugs of its own for a while now althought development seems to now happens again on it after years of a mostly abandonned codebase with a laundry of pull requests ready to be merged with nobody to do it.
> 
> concurrent-ruby and nio4r seems like the best bet for me and a really good choice.

....

**Pull request merged**